verify/verifier.go

```diff
@@ -66,31 +89,44 @@ func NewVerifier(cfg *Config, l log.Logger) (*Verifier, error) {
 	}
 
 	return &Verifier{
-		kzgVerifier: kzgVerifier,
-		verifyCerts: cfg.VerifyCerts,
-		cv:          cv,
+		log:                       log,
+		kzgVerifier:               kzgVerifier,
+		cv:                        cv,
+		rollupBlobInclusionWindow: cfg.RollupBlobInclusionWindow,
 	}, nil
 }
 
 // verifies V0 eigenda certificate type
-func (v *Verifier) VerifyCert(ctx context.Context, cert *Certificate) error {
-	if !v.verifyCerts {
-		return nil
-	}
-
+func (v *Verifier) VerifyCert(ctx context.Context, cert *Certificate, args common.VerifyArgs) error {
 	// 1 - verify batch in the cert is confirmed onchain
 	err := v.cv.verifyBatchConfirmedOnChain(ctx, cert.Proof().GetBatchId(), cert.Proof().GetBatchMetadata())
 	if err != nil {
 		return fmt.Errorf("failed to verify batch: %w", err)
 	}
 
-	// 2 - verify merkle inclusion proof
+	// 2 - verify that the blob cert was submitted to the rollup's batch inbox within the allowed RollupBlobInclusionWindow window.
+	// This is to prevent timing attacks where a rollup batcher could try to game the fraud proof window by including an old DA blob that is about to expire on the DA layer
+	// and is hence not retrievable.
+	if args.RollupL1InclusionBlockNum >= 0 && v.rollupBlobInclusionWindow > 0 {
+		blockNumEigenDA := int64(cert.BlobVerificationProof.BatchMetadata.BatchHeader.ReferenceBlockNumber)
+		blockNumRollup := args.RollupL1InclusionBlockNum
+		// We need blockNumEigenDA < blockNumRollup < blockNumEigenDA + rollupBlobInclusionWindow
+		if !(blockNumEigenDA < blockNumRollup) {
+			return fmt.Errorf("eigenda batch reference block number (%d) needs to be < rollup inclusion block number (%d)", blockNumEigenDA, blockNumRollup)
+		}
+		if !(blockNumRollup < blockNumEigenDA+int64(v.rollupBlobInclusionWindow)) {
```

Comment:
I believe you have an off-by-one error here

The comment in the config struct above states that the following constitutes a stale batch:

```
batch.RBN + RollupBlobInclusionWindow < cert.L1InclusionBlock
```

but it seems what's actually implemented is this:

```
batch.RBN + RollupBlobInclusionWindow <= cert.L1InclusionBlock
```

It might help to have consistent variable names between the two.

Reply:
I think it was doing < everywhere? But I agree it should be <=, so changed. Also updates comments and made naming consistent: [72451c6](https://github.com/Layr-Labs/eigenda-proxy/pull/226/commits/72451c6f3d649c2da206639b028d5d9737d9abd8)

Reply:
The discrepancy was in the config doc:

```
// If batch.RBN + RollupBlobInclusionWindow < cert.L1InclusionBlock, the batch is considered stale and verification will fail.
```

Since this doc is defining what constitutes _staleness_, the implied invariant would be the _inverse_

```
cert.L1InclusionBlock <= batch.RBN + RollupBlobInclusionWindow
```

Your most recent changeset has things correct, since you modified the invariant, but left the staleness definition in the `Config` struct unchanged. Perhaps worth considering stating only the invariant, or only the definition of staleness, to decrease the likelihood for such discrepancies.

Reply:
Don't think there's an issue with that comment.

The invariant to respect is 
```
cert.L1InclusionBlock <= batch.RBN + RollupBlobInclusionWindow
```
but so naturally if that is not true that
```
cert.L1InclusionBlock > batch.RBN + RollupBlobInclusionWindow
```
is true, then the batch is considered stale.